assets/src/stories-editor/blocks/amp-story-text/edit.js

```diff
@@ -25,73 +24,82 @@ import { getBackgroundColorWithOpacity } from '../../../common/helpers';
-			width,
-			content,
-			ampFitText,
-			ampFontFamily,
-		} = attributes;
-
-		// If the block was unselected, make sure that it's not editing anymore.
-		if ( ! isSelected && prevProps.isSelected ) {
-			this.toggleIsEditing( false );
-			this.toggleOverlay( true );
+const TextBlockEdit = ( props ) => {
+	const {
+		attributes,
+		setAttributes,
+		className,
+		clientId,
+		fontSize,
+		isPartOfMultiSelection,
+		isSelected,
+		backgroundColor,
+		customBackgroundColor,
+		textColor,
+		name,
+	} = props;
+	const {
+		placeholder,
+		content,
+		align,
+		ampFontFamily,
+		ampFitText,
+		autoFontSize,
+		width,
+		height,
+		opacity,
+	} = attributes;
+
+	const [ isEditing, setIsEditing ] = useState( false );
+	const [ hasOverlay, setHasOverlay ] = useState( true );
+
+	const toggleIsEditing = ( enable ) => {
+		if ( enable !== isEditing ) {
+			setIsEditing( ! isEditing );
 		}
+	};
 
-		const checkFontSize = ampFitText && (
-			prevProps.attributes.ampFitText !== ampFitText ||
-			prevProps.attributes.ampFontFamily !== ampFontFamily ||
-			prevProps.attributes.width !== width ||
-			prevProps.attributes.height !== height ||
-			prevProps.attributes.content !== content
-		);
+	const toggleOverlay = ( add ) => {
+		if ( add !== hasOverlay ) {
+			setHasOverlay( ! hasOverlay );
+		}
+	};
 
-		if ( checkFontSize ) {
-			maybeUpdateFontSize( this.props );
+	useEffect( () => {
+		if ( isEditing ) {
+			setInputSelectionToEnd( '.is-selected .wp-block-amp-amp-story-text' );
 		}
+	}, [ isEditing ] );
 
-		const checkBlockDimensions = ! ampFitText && (
-			! isEqual( prevProps.fontSize, fontSize ) ||
-			prevProps.attributes.ampFitText !== ampFitText ||
-			prevProps.attributes.ampFontFamily !== ampFontFamily ||
-			prevProps.attributes.content !== content
-		);
+	useEffect( () => {
+		if ( ampFitText ) {
+			maybeUpdateFontSize( props );
+		}
+	}, [ ampFitText, ampFontFamily, width, height, content ] );
 
-		if ( checkBlockDimensions ) {
-			maybeUpdateBlockDimensions( this.props );
+	useEffect( () => {
+		if ( ! ampFitText ) {
+			maybeUpdateBlockDimensions( props );
 		}
+	}, [ ampFitText, fontSize, ampFitText, ampFontFamily, content ] );
 
-		// If the state changed to editing, focus on the text.
-		if ( this.state.isEditing && ! prevState.isEditing ) {
-			setInputSelectionToEnd( '.is-selected .wp-block-amp-amp-story-text' );
+	useEffect( () => {
+		// If the block was unselected, make sure that it's not editing anymore.
+		if ( ! isSelected ) {
+			toggleIsEditing( false );
+			toggleOverlay( true );
 		}
-	}
+	}, [ isSelected ] );
 
-	onReplace = ( blocks ) => {
-		const { attributes, onReplace, name } = this.props;
+	const onReplace = ( blocks ) => {
 		// Make sure that 'undefined' values aren't passed into onReplace.
 		blocks = blocks.filter( ( block ) => 'undefined' !== typeof block );
+
 		if ( ! blocks.length ) {
 			return;
 		}
-		onReplace( blocks.map( ( block, index ) => (
+
+		props.onReplace( blocks.map( ( block, index ) => (
```

Comment:
Shouldn't this be recursion? `props.onReplace` doesn't exist, does it?

Reply:
Nope. A block's edit component gets passed a `onReplace` prop.

See https://github.com/WordPress/gutenberg/blob/fea63772b6e2f71e7422c999e436c44bd3914021/packages/block-editor/src/components/block-list/block.js#L459-L470